### ISSUE TITLE
Unable to Make Local Connection on Windows

### ISSUE THREAD

**Issue by kevinjfoley** [sassoftware/saspy#24]
I get the following error:

> The application could not find a command to launch a SAS Workspace Server.
> SAS process has terminated unexpectedly. RC from wait was: 4294967290
> SAS Connection failed. No connection established. Double check you settings in sascfg.py file.
> 
> Attempted to run program java with the following parameters:['java', '-classpath', 'C:\\Program Files\\SASHOME\\SASDeploymentManager\\9.4\\products\\deploywiz__94360__prt__xx__sp0__1\\deploywiz\\sas.svc.connection.jar;C:\\Program Files\\SASHome\\SASDeploymentManager\\9.4\\products\\deploywiz__94360__prt__xx__sp0__1\\deploywiz\\log4j.jar;C:\\Program Files\\SASHome\\SASDeploymentManager\\9.4\\products\\deploywiz__94360__prt__xx__sp0__1\\deploywiz\\sas.security.sspi.jar;C:\\Program Files\\SASHome\\SASDeploymentManager\\9.4\\products\\deploywiz__94360__prt__xx__sp0__1\\deploywiz\\sas.core.jar;C:\\Users\\user_name\\AppData\\Local\\Continuum\\Anaconda3\\Lib\\site-packages\\saspy\\java\\saspyiom.jar;C:\\Program Files\\SASHOME\\SASFoundation\\9.4\\core\\sasext', 'pyiom.saspy2j', '-host', 'localhost', '-stdinport', '64602', '-stdoutport', '64603', '-stderrport', '64604', '-zero', '']
> 
> Be sure the path to sspiauth.dll is in your System PATH
> 
> No SAS process attached. SAS process has terminated unexpectedly.
> No SAS process attached. SAS process has terminated unexpectedly.
> No SAS process attached. SAS process has terminated unexpectedly.

I've tried adding the path to sspiauth.dll to both my System path variable and the classpath as can be seen above.

**Comment by tomweber-sas** [comment]
The error from trying to make the IOM connection is:
The application could not find a command to launch a SAS Workspace Server.

What version of SAS do you have?

 It appears that the classpath is good; it's starting java and running the saspyiom.jar code which is trying to make the local connection to SAS, but getting that error for some reason. 
I don't believe it's an issue with the sspiauth.dll (yet), only because I think this error is happening before that would be an issue. Also, there a different error that IOM would give if it couldn't find sspiauth.dll.
sspiauth.dll does have to be in the system path, having it on the classpath won't work, as it's a system executable, not a java one. But again, this is a different error currently.

**Comment by kfoley15** [comment]
I've got 9.4 (TS1M1)

**Comment by tomweber-sas** [comment]
Ok, that version sounds good. I'm checking with folks who support this type of IOM connection (out till monday). But, I did get one suggestion:

I suspect it depends on SAS being registered properly in the Windows registry…  that would be the next place to look, but I’m not sure what to check.  Can you double-click a .sas file and get SAS by default?

 Can you confirm that SAS is found when opening a .sas file, and is it's found by running 'sas' from a cmd prompt? Maybe it's not in you path such that it can be found?

**Comment by kfoley15** [comment]
I've always had issues with opening SAS programs from explorer so I've set them to open with a text editor and simply drag them into SAS when I need to run them interactively.

The path to sas.exe wasn't in my Path variable. I've added it and confirmed that `sas` from the command prompt launches SAS.  However even after restarting the python process there I still get the same error.

**Comment by FriedEgg** [comment]
You should also probably check that you have SAS Integration Technologies installed on your machine you can then try to validate a connection outside of SASPy.  I would recommend doing this using PowerShell, but I have an example using Python as well.

https://gist.github.com/FriedEgg/9085229#file-sasworkspaceexample-local-ps1

**Comment by tomweber-sas** [comment]
Ok, I'll have to catch up with the developer on Monday to see what he thinks. I do know others who set this up and it's working for them; obviously it works for me. It's possible the issue with starting your SAS from explorer and running it like this have something on common, but not necessarily, as they really are being driven two different ways.

**Comment by tomweber-sas** [comment]
FriedEgg, that's a good idea. This type of connection is called 'Zero Configuration'; http://support.sas.com/documentation/cdl/en/itechjcdg/64887/HTML/default/viewer.htm#p10vc1l6qy9b9in1jio3czljorpy.htm
There's no object spawner, metadata server or anything else. Does your sample work with that/verify that? This is a little different than the how the IOM access method works when connecting to an object spawner.
Thanks for helping :)

**Comment by FriedEgg** [comment]
@tomweber-sas Yes, there isn't really a direct mapping between the Java and COM implementations but the powershell example I shared is for a local session and essentially matches the Zero Configuration class in Java used by this project.  The Python example in the same gist is for a remote session with object spawner, metadata server, etc...

**Comment by tomweber-sas** [comment]
@FriedEgg Cool. That's great. Thanks!

**Comment by kfoley15** [comment]
I'm able to run the PS script with no issues. 

Also @FriedEgg it looks like your python code might have a typo, I believe [line 12](https://gist.github.com/FriedEgg/9085229#file-sasworkspaceexample-py-L12) is meant to be using the Dispatch method.

**Comment by FriedEgg** [comment]
@kfoley15 thanks for catching that.  I updated that gist.

Is Java in your system path?

**Comment by kfoley15** [comment]
No problem.

It should be, `where java` returns:
>C:\ProgramData\Oracle\Java\javapath\java.exe
>C:\Windows\SysWOW64\java.exe

**Comment by tomweber-sas** [comment]
And if it wasn't finding java, it would be failing long before trying to start SAS, and with a different error message. If you change your class path, to make it invalid, you should see an error that confirms this. Try commenting out the iom jars, leaving saspyiom.jar. That way you should get a java traceback about some iom class not found. That will prove java is starting up, running my main class which then fails trying to load whatever first iom class it loads.

**Comment by FriedEgg** [comment]
I added another example gist of a minimalist test for the ZeroConfigWorkspaceServer using sspi in Java

https://gist.github.com/FriedEgg/9085229#file-sasworkspaceexample-java

```
java "-Djava.library.path=C:\Program Files\SASHome\SASFoundation\9.4\core\sasext" -classpath "C:\Program Files\SASHOME\SASDeploymentManager\9.4\products\deploywiz__94360__prt__xx__sp0__1\deploywiz\sas.svc.connection.jar;C:\Program Files\SASHome\SASDeploymentManager\9.4\products\deploywiz__94360__prt__xx__sp0__1\deploywiz\log4j.jar;C:\Program Files\SASHome\SASDeploymentManager\9.4\products\deploywiz__94360__prt__xx__sp0__1\deploywiz\sas.security.sspi.jar;C:\Program Files\SASHome\SASDeploymentManager\9.4\products\deploywiz__94360__prt__xx__sp0__1\deploywiz\sas.core.jar" SASWorkspaceExample
```

**Comment by FriedEgg** [comment]
This actually raises another point, regarding sspiauth.dll, I see the path `C:\Program Files\SASHome\SASFoundation\9.4\core\sasext` in the original post listed in the Java classpath, but it should be in the java.library.path system property.  SASPy doesn't seem to offer a way to set this, so it is expected to be configured or copied to another path outside of this package.  I don't think it is really relevant to this issue, because that file missing should generate a different error message, but it seems like something in the SASPy configuration that could be updated.

**Comment by tomweber-sas** [comment]
That path (to sspiauth.dll) should just be in the system path variable. It shouldn't be in the class path. Again, I don't think we're getting far enough along to even be looking for that; it is a different error message when that's not found.

**Comment by kfoley15** [comment]
Yeah, I included it in the class path by mistake but it's also in my system path variable.  So are there any next steps on troubleshooting?  Or wait to hear back from the IOM folks next week?

**Comment by tomweber-sas** [comment]
Ok, I have some info that should help. It is possible that the problems you have trying to start SAS from clicking a .sas file might be the same issue. Turns out it's looking in the registry for the command to start SAS. That should've been something set up at installation time. I can't speak to that, but here's what to try:

Execute the following command at a CMD prompt and check the output:
reg query "HKEY_CLASSES_ROOT\CLSID\{440196D4-90F0-11D0-9F41-00A024BB830C}\LocalServer32" /ve

(you can also just use regedit and go look at what's there)

I get the following from that:
HKEY_CLASSES_ROOT\CLSID\{440196D4-90F0-11D0-9F41-00A024BB830C}\LocalServer32
    (Default)    REG_SZ    C:\PROGRA~1\SASHome\SASFOU~1\9.4\SAS.EXE -config C:\PROGRA~1\SASHome\SASFOU~1\9.4\sasv9.cfg -objectserver -nologo -noterminal -noprngetlist

I'm expecting that you'll get an error or something about not finding that (let's see).

Apparently you can run the following to fix that:
[path to sas.exe]\sas.exe --regserver

Can you try this and let me know what have?

**Comment by kfoley15** [comment]
Hmm, I'm able to find that in the registry with the same value you have (I think the tildas may have messed up the formatting a bit)

Also I think it should be `HKEY_CLASSES_ROOT\CLSID\{440196D4-90F0-11D0-9F41-00A024BB830C}\LocalServer32` (extra slash) and `[path to sas.exe]\sas.exe /regserver` just in case anyone else comes across this

**Comment by tomweber-sas** [comment]
I think it's time to engage Tech Support on this. It seems somehow specific to your install; many others are using this w/out issue. I don't know anything else to try, so let's try this path. Can you open a ticket with TS?      This will be a bit of a new experience for them, though we've worked together on having a plan in place for this. You can reference this issue and tell them I asked you to open a ticket. That way we can get right to the folks we need to figure this out.
Thanks,
Tom

**Comment by tomweber-sas** [comment]
@kfoley15 I've engaged out Tech Support team and they found a likely possibility for why you're having this problem. Here's their thoughts:

So everything I’m seeing [in out tracking system] on that error points to the client being unable to access the registry because of a group policy “Prevent access to registry editing tools”. Looks like it comes up pretty often with Forecast Studio for Desktop. If that isn’t set, he’ll probably want to just reinstall Foundation SAS on his host machine to make certain the registry entries are in there correctly.

Can you check to see if this could be the problem? 
Thanks,
Tom

**Comment by kfoley15** [comment]
I believe we do have that group policy enabled.  However my registry seems to be correct based on the info you posted, so I'm not sure why it would need to be edited?

Given that we have that option enabled should I open a ticket with support and work it out with them?  Thanks for all of your help on this so far.

**Comment by tomweber-sas** [comment]
Yes, let's open a track since they have many more resources to help track down things like this. Be aware that this is a whole new thing for our TS dept, so they might not quite know what to do with this, depending upon who gets the call. Just tell them you have already have an issue on github (#24) and that I asked you to open a ticket. That way I can get up directly with them and the three of us can coordinate.
Thanks! 
Tom

**Comment by tomweber-sas** [comment]
@kfoley15  have you had any more luck with this? Did you try changing the permissions, per Tech Support? Did it make a difference? Have you been working w/ TS on the track to see what else it could be?

**Comment by kfoley15** [comment]
Thanks for following up @tomweber-sas - there does seem to be an issue with that group policy, however I started getting another error after changing that which I haven't had the time to troubleshoot, probably safe to close this issue for now and I'll look into the other issue when I get the time

**Comment by tomweber-sas** [comment]
ok @kfoley15 . I'm bummed that we didn't get it working yet, but when you have time again, get back up here and w/ TS so we can get you working. Maybe reinstalling SAS on the PC will fix the registry issues, if that's an option.
I'll close this for now.
Thanks,
Tom

**Comment by tomweber-sas** [comment]
Hey, we now have a workaround for this; see #58 
The code is at master, but I still need to get this into the doc. For now jusr see #58 though, as it's just another key on the config definition. Easy to add.

Thanks,
Tom

**Comment by kevinjfoley** [comment]
Looks like it's no working, FYI for anyone else I installed the latest version from github using `pip install https://github.com/sassoftware/saspy/tarball/master`

Thanks for following up on this!

**Comment by tomweber-sas** [comment]
Is that 'now working', or 'not working'? Did you add the javaparms key with the correct sas.exe command for your install?

**Comment by kevinjfoley** [comment]
Sorry, it should be 'now working'

**Comment by tomweber-sas** [comment]
Sweet! You had me worried there for a second :)
